RFC: clean up `map` and `collect` with iterator traits

JeffBezanson opened the pull request:
This is a proposal for cleaning up our iteration, mapping, and collecting. Between `copy[!]`, `collect`, `map`, and comprehensions we had subtly different behaviors for similar operations. We also had some repeated code and inefficient generic fallbacks.

This change replaces all that with `collect` of various iterators. There are four core algorithms for this, based on whether an iterator has a known length/shape and whether it has a known element type. To select the right algorithm, I added some HolyTraits for iterators. I seem to recall iterator traits being discussed before, but I can't find a reference.

The biggest immediate benefit is that generic `map` is faster and better-typed.
Before:

```
julia> @time map(tuple, 11:2000000, enumerate(11:2000000));
  2.745751 seconds (18.00 M allocations: 505.257 MB, 29.04% gc time)
```

giving a `Vector{Any}`.
After:

```
julia> @time map(tuple, 11:2000000, enumerate(11:2000000));
  0.570696 seconds (6.00 M allocations: 198.356 MB, 11.51% gc time)
```

giving a `Vector{Tuple{Int64,Tuple{Int64,Int64}}}`.
This is achieved with the following glorious definition(s):

```
spread(f) = (args)->f(args...)
map(f, iters...) = collect(Generator(spread(f),zip(iters...)))
```

This definition is _almost_ as fast as existing array functions:

```
julia> A = rand(1000,1000);

julia> @time A+A;
  0.001964 seconds (9 allocations: 7.630 MB)

julia> @time map(+,A,A);
  0.004792 seconds (11 allocations: 7.630 MB)
```

Hopefully this gap can be closed.

A major issue here is iterator shape preservation. Some iterators just wrap others and so should obviously preserve their shape, like `Generator`. Others are less obvious. Since zipped iteration is so common when working with arrays, I made `zip` shape-preserving. `product` is currently 1d only, but could for example concatenate the shapes of its component iterators.

StefanKarpinski committed:
ENV: test errors for non-existent keys, NUL-containing keys

StefanKarpinski committed:
ENV: cosmetic changes

StefanKarpinski committed:
ENV: inline _jl_win_getenv into its only call site

StefanKarpinski committed:
ENV: move system-specific code into large os-specific blocks

Commit:
added os_only documentation

StefanKarpinski committed:
ENV: change access_env from macro to higher-order function

KristofferC committed:
add specialized copy! for sparse matrices

when the matrices have the same size we can get away with simple copying of the internal fields.

pkofod committed:
Added tests to increase coverage in base/rationals.jl

andreasnoack committed:
Reenable test of condition number estimates of sparse matrices
and adjurst the tolerance. Fixes #14778

JeffBezanson committed:
a few more micro-optimizations to parsing

hey, every bit helps.

JeffBezanson committed:
make `map` and `collect` more general and uniform by adding iterator traits

Cody-G committed:
patches #13682 to fix memory mapping of large files on Windows

tbreloff commented:
+1. Does this mean we could have a correctly typed unzip as well?

On Wednesday, February 17, 2016, Jeff Bezanson <notifications@github.com>
wrote:

> This is a proposal for cleaning up our iteration, mapping, and collecting.
> Between copy[!], collect, map, and comprehensions we had subtly different
> behaviors for similar operations. We also had some repeated code and
> inefficient generic fallbacks.
>
> This change replaces all that with collect of various iterators. There
> are four core algorithms for this, based on whether an iterator has a known
> length/shape and whether it has a known element type. To select the right
> algorithm, I added some HolyTraits for iterators. I seem to recall iterator
> traits being discussed before, but I can't find a reference.
>
> The biggest immediate benefit is that generic map is faster and
> better-typed.
> Before:
>
> julia> @time map(tuple, 11:2000000, enumerate(11:2000000));
>   2.745751 seconds (18.00 M allocations: 505.257 MB, 29.04% gc time)
>
> giving a Vector{Any}.
> After:
>
> julia> @time map(tuple, 11:2000000, enumerate(11:2000000));
>   0.570696 seconds (6.00 M allocations: 198.356 MB, 11.51% gc time)
>
> giving a Vector{Tuple{Int64,Tuple{Int64,Int64}}}.
> This is achieved with the following glorious definition(s):
>
> spread(f) = (args)->f(args...)
> map(f, iters...) = collect(Generator(spread(f),zip(iters...)))
>
> This definition is *almost* as fast as existing array functions:
>
> julia> A = rand(1000,1000);
>
> julia> @time A+A;
>   0.001964 seconds (9 allocations: 7.630 MB)
>
> julia> @time map(+,A,A);
>   0.004792 seconds (11 allocations: 7.630 MB)
>
> Hopefully this gap can be closed.
>
> A major issue here is iterator shape preservation. Some iterators just
> wrap others and so should obviously preserve their shape, like Generator.
> Others are less obvious. Since zipped iteration is so common when working
> with arrays, I made zip shape-preserving. product is currently 1d only,
> but could for example concatenate the shapes of its component iterators.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/JuliaLang/julia/pull/15123
> Commit Summary
>
>    - make `map` and `collect` more general and uniform by adding iterator
>    traits
>
> File Changes
>
>    - *M* base/abstractarray.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-0> (150)
>    - *M* base/array.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-1> (99)
>    - *M* base/dict.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-2> (3)
>    - *M* base/generator.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-3> (39)
>    - *M* base/inference.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-4> (2)
>    - *M* base/iterator.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-5> (46)
>    - *M* test/abstractarray.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-6> (14)
>    - *M* test/arrayops.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-7> (2)
>    - *M* test/datafmt.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-8> (2)
>    - *M* test/unicode/utf8proc.jl
>    <https://github.com/JuliaLang/julia/pull/15123/files#diff-9> (2)
>
> Patch Links:
>
>    - https://github.com/JuliaLang/julia/pull/15123.patch
>    - https://github.com/JuliaLang/julia/pull/15123.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/15123>.
>

StefanKarpinski committed:
Merge pull request #15122 from JuliaLang/sk/envtweaks

some tweaks to code for the EnvHash type

JeffBezanson committed:
Merge pull request #15124 from JuliaLang/jb/parsefaster2

a few more micro-optimizations to parsing

tkelman committed:
Merge pull request #15125 from Cody-G/cjg/fix_mmap_win

patches #13682 to fix memory mapping of large files on Windows

helgee committed:
Clarify section about Base.convert fall-back.

helgee committed:
Remove trailing whitespace.

mbauman committed:
Merge pull request #15104 from KristofferC/kc/sparse_copy

add specialized copy! for sparse matrices

justbur committed:
julia-mode: Fix prompt regexp

Fixes #14002

The primary problem was the lack of a space at the end of the regexp,
which didn't match the output of the shell process (does have a space).
This meant that emacs sometimes placed the point one character before
the end of the buffer, which prevented M-p from working.

This new regexp also allows for other prompts, such as "shell> "

StefanKarpinski committed:
mmap: remove unnecessary use of utf16

JeffBezanson committed:
Merge pull request #15132 from helgee/master

[WIP] Clarify section about Base.convert fall-back.

yuyichao committed:
Merge pull request #15135 from justbur/julia-mode-fix-prompt

julia-mode: Add space to prompt regexp

tkelman committed:
Merge pull request #15108 from morris25/add_os_only_documentation

RFC: added os_only documentation

andreasnoack committed:
Merge pull request #15113 from JuliaLang/anj/cond

Reenable test of condition number estimates of sparse matrices

StefanKarpinski committed:
Merge pull request #15105 from pkofod/master

Added tests to increase coverage in base/rationals.jl

StefanKarpinski committed:
Merge pull request #15137 from JuliaLang/sk/rm-mmap-utf16

mmap: remove unnecessary use of utf16

JeffBezanson committed:
Merge pull request #14848 from JuliaLang/jb/generators

RFC: generator expressions
